openmc/model/triso.py

```diff
@@ -153,3 +162,843 @@ def create_triso_lattice(trisos, lower_left, pitch, shape, background):
+    particles apart along the line joining their centers and the outer diameter
+    is decreased. Iterations continue until the two diameters converge or until
+    the desired pf is reached.
+
+    References
+    ----------
+    .. [1] W. S. Jodrey and E. M. Tory, "Computer simulation of close random
+       packing of equal spheres", Phys. Rev. A 32 (1985) 2347-2351.
+
+    """
+
+    def get_domain_volume():
+        """Calculates the volume of the container in which the TRISO particles
+        are packed.
+
+        Returns
+        -------
+        float
+            Volume of the domain.
+
+        """
+
+        if domain_shape is 'cube':
+            return domain_length**3
+        elif domain_shape is 'cylinder':
+            return domain_length * pi * domain_radius**2
+        elif domain_shape is 'sphere':
+            return 4/3 * pi * domain_radius**3
+
+
+    def get_cell_length(radius):
+        """Calculates the length of a lattice element in x-, y-, and
+        z-directions.
+
+        Parameters
+        ----------
+        radius : float
+            Radius of the particle.
+
+        Returns
+        -------
+        tuple of float
+            Length of lattice cell in x-, y-, and z-directions.
+
+        """
+
+        if domain_length:
+            m = domain_length/int(domain_length/(4*radius))
+        if domain_radius:
+            n = 2*domain_radius/int(domain_radius/(2*radius))
+
+        if domain_shape is 'cube':
+            return (m, m, m)
+        elif domain_shape is 'cylinder':
+            return (n, n, m)
+        elif domain_shape is 'sphere':
+            return (n, n, n)
+
+
+    def get_boundary_extremes():
+        """Calculates the minimum and maximum positions in x-, y-, and
+        z-directions where a particle center can be placed within the domain.
+
+        Returns
+        -------
+        llim, ulim : tuple of float
+            Minimum and maximum position in x-, y-, and z-directions where
+            particle center can be placed.
+
+        """
+
+        if domain_length:
+            x_min = radius
+            x_max = domain_length - radius
+        if domain_radius:
+            r_min = radius - domain_radius
+            r_max = domain_radius - radius
+
+        if domain_shape is 'cube':
+            return (x_min, x_min, x_min), (x_max, x_max, x_max)
+        elif domain_shape is 'cylinder':
+            return (r_min, r_min, x_min), (r_max, r_max, x_max)
+        elif domain_shape is 'sphere':
+            return (r_min, r_min, r_min), (r_max, r_max, r_max)
+
+
+    def get_particle_offset():
+        """Calculates the offset in x-, y-, and z-directions of the particle
+        center based on the domain center
+
+        Returns
+        -------
+        tuple of float
+            Amount to offset particle center in x-, y-, and z-directions
+
+        """
+
+        if domain_shape is 'cube':
+            return np.array(domain_center) - 3*(domain_length/2,)
```

Comment:
`np.array(domain_center) - domain_length/2` will work just as well because numpy automatically broadcasts the scalar.